File: java/client/src/main/java/glide/api/models/commands/geospatial/GeoSearchResultOptions.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide.api.models.commands.geospatial;

import glide.api.commands.GeospatialIndicesBaseCommands;
import glide.api.models.commands.SortOrder;
import java.util.ArrayList;
import java.util.List;

/**
 * Optional arguments for {@link GeospatialIndicesBaseCommands#geosearch(String,
 * GeoSearchOrigin.SearchOrigin, GeoSearchShape, GeoSearchOptions)} command that contains up to 3
 * optional input, including:
 *
 * <ul>
 *   <li>{@link SortOrder} to order the search results by the distance to the center point of the
 *       search area.
 *   <li><code>COUNT</code> to limits the number of search results.
 *   <li><code>ANY</code> option, which could be used with <code>COUNT</code> only, makes command to
 *       return as soon as enough matches are found. This means that the results might may not be
 *       the ones closest to the origin.
 * </ul>
 *
 * @see <a href="https://valkey.io/commands/geosearch/">valkey.io</a>
 */
public class GeoSearchResultOptions {
    /** Valkey API keyword used to perform geosearch with count. */
    public static final String COUNT_VALKEY_API = "COUNT";

    /**
     * Valkey API keyword used to change search behavior to return as soon as enough matches are
     * found.
     */
    public static final String ANY_VALKEY_API = "ANY";

    /** Indicates the order the result should be sorted in. See {@link SortOrder} for detail. */
    private final SortOrder sortOrder;

    /** Indicates the number of matches the result should be limited to. */
    private final long count;

    /** Whether to allow returning as soon as enough matches are found. */
    private final boolean isAny;

    /** Define number of search results. */
    public GeoSearchResultOptions(long count) {
        this.sortOrder = null;
        this.count = count;
        this.isAny = false;
    }

    /** Define number of search results <code>ANY</code> option. */
    public GeoSearchResultOptions(long count, boolean isAny) {
        this.sortOrder = null;
        this.count = count;
        this.isAny = isAny;
    }

    /** Define the sort order only. */
    public GeoSearchResultOptions(SortOrder order) {
        this.sortOrder = order;
        this.count = -1;
        this.isAny = false;
    }

    /** Define the sort order and count. */
    public GeoSearchResultOptions(SortOrder order, long count) {
        this.sortOrder = order;
        this.count = count;
        this.isAny = false;
    }

    /** Configure all parameters. */
    public GeoSearchResultOptions(SortOrder order, long count, boolean isAny) {
        this.sortOrder = order;
        this.count = count;
        this.isAny = isAny;
    }

    /**
     * Converts GeoSearchResultOptions into a String[].
     *
     * @return String[]
     */
    public String[] toArgs() {
        List<String> arguments = new ArrayList<>();

        if (count > 0) {
            arguments.add(COUNT_VALKEY_API);
            arguments.add(Long.toString(count));

            if (isAny) {
                arguments.add(ANY_VALKEY_API);
            }
        }

        if (sortOrder != null) {
            arguments.add(sortOrder.toString());
        }

        return arguments.toArray(new String[0]);
    }
}

File: node/src/commands/geospatial/GeoSearchResultOptions.ts
/**
 * Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0
 */

// Import below added to fix up the TSdoc link, but eslint blames for unused import.
/* eslint-disable-next-line @typescript-eslint/no-unused-vars */
import { BaseClient } from "src/BaseClient";

/** Defines the sort order for nested results. */
export enum SortOrder {
    /** Sort by ascending order. */
    ASC = "ASC",
    /** Sort by descending order. */
    DESC = "DESC",
}

// This class unites `GeoSearchResultOptions` and `GeoSearchOptions` of the java client
/**
 * Optional parameters for {@link BaseClient.geosearch|geosearch} command which defines what should be included in the
 * search results and how results should be ordered and limited.
 */
export class GeoSearchResultOptions {
    private readonly withCoord: boolean;
    private readonly withDist: boolean;
    private readonly withHash: boolean;
    private readonly sortOrder?: SortOrder;
    private readonly count?: number;
    private readonly isAny: boolean;

    /**
     * Optional parameters for {@link BaseClient.geosearch|geosearch} command which defines what should be included in the
     * search results and how results should be ordered and limited.
     *
     * @param withCoord - Include the coordinate of the returned items.
     * @param withDist - Include the distance of the returned items from the specified center point.
     *     The distance is returned in the same unit as specified for the `searchBy` argument.
     * @param withHash - Include the geohash of the returned items.
     * @param sortOrder - Indicates the order the result should be sorted in.
     * @param count - Indicates the number of matches the result should be limited to.
     * @param isAny - Whether to allow returning as enough matches are found. This requires `count` parameter to be set.
     */
    public constructor(options: {
        withCoord?: boolean;
        withDist?: boolean;
        withHash?: boolean;
        sortOrder?: SortOrder;
        count?: number;
        isAny?: boolean;
    }) {
        this.withCoord = options.withCoord ?? false;
        this.withDist = options.withDist ?? false;
        this.withHash = options.withHash ?? false;
        this.sortOrder = options.sortOrder;
        this.count = options.count;
        this.isAny = options.isAny ?? false;
    }

    /** Convert to the command arguments according to the Valkey API. */
    public toArgs(): string[] {
        const args: string[] = [];

        if (this.withCoord) args.push("WITHCOORD");
        if (this.withDist) args.push("WITHDIST");
        if (this.withHash) args.push("WITHHASH");

        if (this.count) {
            args.push("COUNT", this.count?.toString());

            if (this.isAny) args.push("ANY");
        }

        if (this.sortOrder) args.push(this.sortOrder);

        return args;
    }
}
